homeassistant/components/hue/hue_sensor.py

```diff
@@ -0,0 +1,237 @@
+
+    if not bridge.available:
+        _LOGGER.info('Reconnected to bridge %s', bridge.host)
+        bridge.available = True
+
+    new_sensors = []
+    sensor_device_names = {}
+
+    # Physical Hue motion sensors present as three sensors in the API: a
+    # presence sensor, a temperature sensor, and a light level sensor. Of
+    # these, only the presence sensor is assigned the user-friendly name that
+    # the user has given to the device. Each of these sensors is linked by a
+    # common device_id, which is the first twenty-three characters of the
+    # unique id (then followed by a hyphen and an ID specific to the individual
+    # sensor).
+    #
+    # To set up neat values, and assign the sensor entities to the same device,
+    # we first, iterate over all the sensors and find the Hue presence sensors,
+    # then iterate over all the remaining sensors - finding the remaining ones
+    # that may or may not be related to the presence sensors.
+    for item_id in api:
+        if item_id in current:
+            continue
+
+        name = PRESENCE_NAME_FORMAT.format(api[item_id].name)
+        if api[item_id].type == aiohue.sensors.TYPE_ZLL_PRESENCE:
+            sensor = HuePresence(api[item_id], name, bridge)
+            sensor_device_names[sensor.device_id] = api[item_id]
+            current[item_id] = sensor
+            if binary:
+                new_sensors.append(sensor)
+
+    # Iterate again now we have all the presence sensors, and add the related
+    # sensors with nice names.
+    for item_id in api:
+        if item_id in current:
+            continue
+
+        # Work out the shared device ID, as described above
+        device_id = api[item_id].uniqueid
+        if device_id and len(device_id) > 23:
+            device_id = device_id[:23]
+        name = api[item_id].name
+        primary_sensor = None
+        if api[item_id].type == aiohue.sensors.TYPE_ZLL_LIGHTLEVEL:
+            darkness_name = name + ' is not dark'
+            if device_id in sensor_device_names:
+                primary_sensor = sensor_device_names[device_id]
+                name = LIGHT_LEVEL_NAME_FORMAT.format(
+                    primary_sensor.name)
+                darkness_name = IS_DARK_NAME_FORMAT.format(
+                    primary_sensor.name)
+            current[item_id] = HueLightLevel(
+                api[item_id], name, bridge, primary_sensor=primary_sensor)
+            if binary:
+                darkness = HueNotDarkness(
+                    api[item_id], darkness_name, bridge,
+                    primary_sensor=primary_sensor)
+                new_sensors.append(darkness)
+        elif api[item_id].type == aiohue.sensors.TYPE_ZLL_TEMPERATURE:
+            if device_id in sensor_device_names:
+                primary_sensor = sensor_device_names[device_id]
+                name = TEMPERATURE_NAME_FORMAT.format(
+                    primary_sensor.name)
+            current[item_id] = HueTemperature(
```

Comment:
this should check of `binary` ? 

I think that it's easier to extract this convert-things-to-entitites logic into 2 methods, one for binary and one for normal sensosrs.

Reply:
Refactored to avoid this, see what you think of the new setup code.